### ISSUE TITLE
Some hardware to work with your firmware

### ISSUE THREAD

**Issue by raavhimself** [TheZeroBeast/MM-control-01#58]
https://www.thingiverse.com/thing:3245715

**Comment by TheZeroBeast** [comment]
Hell Yeah 👏 that looks very good, can you please let me know how it goes?

**Comment by TheZeroBeast** [comment]
How goes the prints with your extruder mod @raavhimself?

**Comment by raavhimself** [comment]
Prints were really well until the sensor filament died. I am waiting for the replacement, supposedly arriving today.
In any case.I am now two steps beyond with my hardware mod. I completely redesigned the extruder body's upper part and introduced a blade!!! I think PRUSA put the blade in the wrong place. I just thought I should kill the strings where they are created, that is in the extruder.
My new contraption works really, really well. No strings AT ALL. Not even with PETG. The thing works so well that I have been able to make two prints completely unassisted with the standard PRUSA firmware (no filament sensor). The only problem I have found are the globs. As the PRUSA firmware is "blind", a glob can block the ptfe tube beyond the pulleys and there you have a skipped layer. I am eager to continue testing the device with your new firmware.
The thing is that my device works so well that, after some more testing, I intend to place it in GRABCAD for licensing reasons. I must admit that I am a bit resented with PRUSA for selling me a device that just does not work, and making me deviate from my tasks for so long, so I am not giving them my engineering for free. I will apply also a non payment policy for my device, that is: if you do not want to pay for my design, you won't have to. 
I really will be honored if you want to work with the device I have designed. I am including a picture so you can have an idea. I think I will release it in about a week, once I do some more testing. I'll let you know
![the frog](https://user-images.githubusercontent.com/37851425/51103122-e5ae5a00-17e1-11e9-8e5b-1af26ca1ae4b.jpeg)
![thefrog](https://user-images.githubusercontent.com/37851425/51103123-e5ae5a00-17e1-11e9-9d95-706c0f7bd679.jpg)

**Comment by TheZeroBeast** [comment]
Very interesting 🤔 I can see that would be a help for sure, how do you go about using the blade?
I agree that PRUSA has pissed a lot of people off, I have been lost in code for months now just to have my printer do what it should have when we got the MMU2.
Sell and license away.
Feel free to PM me on facebook if you'd prefer to chat about this cool blade idea in private.

**Comment by raavhimself** [comment]
I see that the picture might be difficult to understand. The blade is somewhere inside that green part. Just for your reference, in the upper part of the extruder, where PRUSA design is using 7 parts including the sensor and the bolts, all of them fixed, my design uses something like 28 different parts, and although half of them are moving parts, the thing really works like a swiss clock.

**Comment by TheZeroBeast** [comment]
Very cool, hope it keeps the prints coming for you.
I've gotten around the last few skipped layers by getting Jam detection working so when a tip stuffs a load it picks it up even though the filament sensor did the load right.
You might want to check other sources for FSensors than PRUSA, I got some from Ebay that are better made, delivered quicker and quite a bit cheaper. AU$19 each. I now have a few backup just in case.

I'm running a custom extruder mod I tipped up last week with passthrough which is going great guns with sensor reliability.
<img width="590" alt="screen shot 2019-01-14 at 7 24 23 pm" src="https://user-images.githubusercontent.com/20939095/51104530-22a23d00-1832-11e9-9e43-3e438efaca82.png">
<img width="778" alt="screen shot 2019-01-14 at 7 29 46 pm" src="https://user-images.githubusercontent.com/20939095/51104780-cd1a6000-1832-11e9-9731-a3bfc7c02bed.png">
<img width="790" alt="screen shot 2019-01-14 at 7 29 13 pm" src="https://user-images.githubusercontent.com/20939095/51104783-d4416e00-1832-11e9-85e2-7b796e5f56ad.png">

**Comment by AbeFM** [comment]
OMG isn't the tower of 50 things that each almost work getting old.

I'm a couple weeks into debugging the Skelestruder, and trying to seperate
out the issue of firmware, slicer, extruder, etc.
Especially since some of these things are closed source, it's so annoying
to need to run through a week of reverse engineering for every little hold
and blob. I want to fix up the sensor on this thing but of course I'd have
to do it from scratch.

What is triggering your blade? I'm a little worried because of the
overheating and weight issues of the stock Prusa set up. In that regard,
the Skelestruder is excellent (including filament constraint!).

I think the blade at the top makes more sense, if it goes into the hole,
everything's ok? I think I missed it, please explain the logic - when would
you get your failure and what kind where the lower blade would help?

**Comment by AbeFM** [comment]
Raav: Are you using an indirect sensor? Have you tried a grooved pulley?

I've yet to try this, and I cannot get one to fit easily into the
Skelestruder. Designs are being kept private, so mostly all I can do is
stare, hope, and suggest.

**Comment by raavhimself** [comment]
AbeFM - I'd like to send you a couple of images in a private message to show you the idea I am working in. May I have your mail address? Mine is gcon114@gmail.com 
Summarizing, the filament triggers the blade everytime it retracts from the head. The device does not cut the tip, it just prevents hairs from appearing by presing a blade against a TPU roller blocking what was before the filament's path. If hairs do appear, they are extremely thin and pretty short. They are never an issue. Filament sensor aims at this same TPU roller. The blade mechanism is triggered automatically each time the filament displaces the roller.
I haven`t started tinkering with the settings yet (too many machine breaks, first the PINDA probe, then the sensor) but I guess my device will simplify settings a lot. For example, I haven´t tried multimaterial prints with PETG yet, but when I am printing parts with PETG, filament tips are always as good as they can be, and repeated loads to nozzle from different PETG colors wortk always perfect. No sign of strings. and I haven't touched the PET settings at all. (Prusa PET and generic PET)

About the pulley: I did try the grooved pulley. As a funny fact it costed me about 6€ in aliexpress what prusa sells for something like 49.5 €,. Finally I've come to the conclussion that the TPU roller works much better. There's enough guidance and drag and it looks like less resistance in the loads, that were really noisy with the pulley. Also, it looks like the soft roller has the strange effect of combing the hairs backwards in the loads. 

Regarding the sensor, I'd say my last version works flawlessly. I mean, when asked  to perform, like in a load to nozzle operation, it always works just perfect. Another different thing is while printing. The machine is stopped, but what failed? Was it a sensor reading, the firmware, the roller device, the hardware in the PCB of the sensor itself or perhaps a hair that stopped the roller from rolling? Who knows. In my opinion, the important thing is that the firmware can recover always from a hardware crash and that the hardware crashes must be rare and easy to clear.

Regarding my broken  sensor, nothing burned that I can see. The new sensor is a different version with some new fancy green thing. Looks cool like if it worked better. We'll see.
![sensores](https://user-images.githubusercontent.com/37851425/51171737-f8905f80-18b1-11e9-8b28-fe0e7cdcee94.jpeg)

**Comment by raavhimself** [comment]
What it is clear is that you need a certain measure of patience to operate a MMU2. Five hours into a print and everything is smooth as silk. Then in a freak loop convolution, far away in the spool holders, one filament strangles another... and here we go again for the number eleven hundred attempt on the four color aztec.

**Comment by raavhimself** [comment]
Zero - How important is this "distance from bondtech to sensor" in your firmware?  I  have calculated PRUSA's stock distance and it is 16 mm. In my design it is 31.5 mm. Even though, I haven't touched your hex files ((I am really clumsy) and the thing seems to work just fine with them. I am in the opinion that if it does work you better not fix it, but I might be missing something

**Comment by TheZeroBeast** [comment]
Much and more patience!!!

@raavhimself there is a calibration for this now so no code changes required.

**Comment by raavhimself** [comment]
So it auto calibrates itself? I suppose it has all the data it needs. The bowden tube calibration tells it how far the gear pulleys are and the filament is always going to find the sensor somewhere in the way

**Comment by TheZeroBeast** [comment]
It isn’t automatic calibration, check the readme on how to do it but it accomodates different sensor to bondtech distance

**Comment by raavhimself** [comment]
It works so well without calibration that I am reluctant to do it

**Comment by raavhimself** [comment]
I don´t know if I should open a separate issue but it stalled. I made two successful prints bjut the third failed. Some pauses but the machine recovered well from all, Most of the pauses were due to bad sensor readings (my design traps all the bits and debris there, and that causes misreadings, but this is very convenient because cleaning is easy and with your firmware I get no jams, put pauses).
The device got kind of crazy a few times, for example pausing and flashing the led for no aparent reason, I push the middle button and next load is just ok. The stall: Out of nowhere, the same situation except this time pushing the button I just get four BRRRRTs from the barrel roll and that's it. I guess they are necessary but I hate those BRRRRT
![fail](https://user-images.githubusercontent.com/37851425/51325433-2ff73b80-1a6d-11e9-8396-55bc98012c47.jpeg)

**Comment by raavhimself** [comment]
![whatsapp image 2019-01-18 at 22 46 11](https://user-images.githubusercontent.com/37851425/51414617-5c957b00-1b73-11e9-8b95-e6979e810908.jpeg)
First shy try with PETG. No strings, no globs, no hairs, no dust, just a little bit of fluff in the trap. Looks like it could do a lot of filament changes without cleaning. Default settings

**Comment by TheZeroBeast** [comment]
Hey @raavhimself, can you give the latest release code a try please?

**Comment by TheZeroBeast** [comment]
How goes your frog my friend?

**Comment by raavhimself** [comment]
Hi Zero!!
The frog runs pretty well, I mean, it works as it should. I ran into a
phase of bad sensor readings that ended when I changed the sensor again
(and that makes the fifth time) but I am pretty happy with the device. From
what I have heard from a girl that has two printers, the frog works much
better than the MK3S, but I haven't installed PRUSA's device so I don´t
know first hand. I printed all the parts of the MK3S and so, but finally I
decided not replace something that works all right for something that...
who knows. In any case, the frog does not run with the MK3S firmware, only
with your FW, so the fact is that very few people has tried it. (I really
don't give a shit, I repaired my machine, and that was the main point. The
truth is that I haven't gone back to visit the PRUSA forum again)
In any case I am not doing multicolor printing lately. Summer here in the
northern hemisphery, and in my summer holidays I have built and crashed a
Pietenpol, I lost a drone in an unfortunate chain of events and designed
and built a 3d printed RC hovercraft that has been the terror of the beach.
Great fun!!
